### ISSUE TITLE
Download/Upload KBytes/sec Preferences not working as expected - Burst behavior

### ISSUE THREAD

**Issue by romw** [BOINC/boinc#1066]
**Reported by JacobKlein on 22 Apr 41267638 11:33 UTC**
When testing Download and Upload preferences, on my Windows 7 x64 machine, it does not look like speeds are being limited correctly. Windows Task Manager shows limiting, but the speeds burst above the limit, then go to 0, repeatedly.

If I limit to 40 KB/s, I'd expect a smooth and steady 40 KB/s, not 0 for a few seconds, then 85 briefly, then to 0, then to 75, etc.  Yet, that's the behavior I'm observing in Task Manager, with an update speed of 1 second.

Below are some emails where I originally reported this on the Alpha Mailing list.

===================[Klein jacob_w_klein at msn.com[[BR]([BR]]Jacob)]Wed Mar 16 !22:16:14 PDT 2011[(finally) finished my 6.12.18 alpha preferences testing.[[BR]([BR]][[BR]]I've)]I reported the results in the test results, but figured I'd mention the major bug here also.[whenever I have a KBytes/sec limit, the download rate sometimes just stops completely, and sometimes starts/stops.[[BR]([BR]][[BR]]Essentially,)]It certainly isn't smooth like I would expect.[internet normally downloads at ~330KB/sec, and my testing was a setting of 40 KBytes/sec as the limit.[[BR]([BR]]My)]I was resetting projects to get transfers to kick in.[PC has 2 network adapters as seen in Task Manager.[[BR]([BR]][[BR]]My)]- "Ethernet 2" is what is used for Internet and BOINC.["Bluetooth Network Connection" is never used.[[BR]([BR]]-)]... but I get the same results when I disable the Bluetooth and restart BOINC, so I don't think that adapter is interfering.[kind of weird, because, even though the GUI shows 4 downloads each of 10 KBytes/sec..[[BR]([BR]][[BR]]It's)]Windows Task Manager (which shows the actual data, and is updating every second), shows sporadic jumps above 40, and then 0, and isn't smooth.[almost as if the GUI "Speed" calculation is wrong, but BOINC uses that wrong calculation to limit the speed.[[BR]([BR]]It's)]It could even be a timing problem or race condition, not sure.[anyone replicate the odd behavior?[[BR]([BR]][[BR]]Can)][smaller preference bugs that I found:[[BR]([BR]]Other)]- "Confirm before connecting to internet" did not do anything for me (on Ethernet), but tool tip does not say dial-up only.[If Network is suspended because of "Suspend Network" activity, it should say "Suspended - user request", even if a time-of-day preference is set.[[BR]([BR]]-)]- Day-of-week override does not have tool tips on day hover or check box hover[BOINC Manager inputs needs better sanitization[[BR]([BR]][[BR]]Several)]- "Idle minutes" preference input should not allow negative numbers["Hours" inputs should not allows entries like !13:239 [[BR]([BR]]-)]- "% of the processors" should not allow negative numbers or numbers >100["% CPU Time" should not allow negative numbers or numbers >100[[BR]([BR]]-)]- "While processor usage is less than" should not allow negative numbers or numbers >100["Maximum download rate" should not allow negative numbers[[BR]([BR]]-)]- "Maximum upload rate" should not allow negative numbers[few web-based inputs need better sanitization[[BR]([BR]][[BR]]A)]- "Maximum download rate" should not allow negative numbers["Maximum upload rate" should not allow negative numbers[[BR]([BR]]-)][regards,[[BR]([BR]]Kind)]Jacob Klein[Walton romw at romwnet.org[[BR]([BR]][[BR]]===================[[BR]]Rom)]Thu Mar 17 !11:44:53 PDT 2011[take a look this weekend.[[BR]([BR]][[BR]]I'll)][app only knows what it sends out over the outgoing socket and what[[BR]([BR]]Our)]is read from the incoming socket, things like TCP/IP overhead are not[in the BOINC calculations.[[BR]([BR]]included)][Manager from the Task Manager is displaying overall throughput[[BR]([BR]]Resource)]including TCP/IP's overhead.  Resource Manager also includes the sum[of all traffic across all processes.[[BR]([BR]]total)][far as the input validation in the Manager goes, I'll get that[[BR]([BR]]As)]cleaned up for 7.0.  I'm not going to hold up the 6.12 release for it.[Rom[[BR]([BR]][[BR]]-----)][Klein jacob_w_klein at msn.com[[BR]([BR]]===================[[BR]]Jacob)]Tue Mar 22 !06:00:10 PDT 2011[your investigation find anything?[[BR]([BR]][[BR]]Did)][Walton romw at romwnet.org[[BR]([BR]]===================[[BR]]Rom)]Tue Mar 22 !08:38:09 PDT 2011[believes it is only sending out what it is told to.  The odd stuff[[BR]([BR]][[BR]]BOINC)]is happening at a lower level of the software stack.[Manager is including the TCP/IP overhead while what BOINC[[BR]([BR]][[BR]]Resource)]reports does not.  I don't know if the burst type behavior is Windows or[ If I had to take a guess I would say Windows.  Without knowing[[BR]([BR]]Libcurl.)]how your network is configured, Windows does try to make use of things[Jumbo Frames and TCP checksum offloading to increase network[[BR]([BR]]like)]throughput.[this point I'm happy as long as BOINC doesn't saturate the WAN[[BR]([BR]][[BR]]At)](Internet) connection.[[BR]][[BR]]----- Rom

**Comment by romw** [comment]
**Commented by davea on 6 Feb 41267689 17:20 UTC**
BOINC doesn't enforce bandwidth limits itself;
it relies on the HTTP library (libCURL) to do so.
I guess libCURL does this in a coarse-grained way,
e.g. by turning transmission on and off at 1-second scale.
As long as the average is enforced correctly on the
10+ sec time scale,
I don't think it's worth filing a bug to libCURL about this.

**Comment by romw** [comment]
**Commented by JacobKlein on 21 Jun 41271216 15:06 UTC**
Can we upgrade to the latest version of libCurl? From the best I can tell, we last dropped 7.19.7, but since then, there have been 2 years worth of bugfixes (including ones related to speeds/rates/timers), including 2 vulnerability plugs.

Could we upgrade to 7.21.4 or better, to see if it fixes the problems in this ticket?

**Comment by romw** [comment]
**Commented by Nicolas on 11 Apr 41272457 01:46 UTC**
BOINC on Windows is dynamically linked to libcurl; so ''in theory'' you could replace your `C:\Program Files\BOINC\libcurl.dll` with a newer one, and test yourself if it fixes the problem.

However, it seems the curl website doesn't have Windows 64-bit downloads for libcurl, only 32-bit... `:/`

**Comment by romw** [comment]
**Commented by JacobKlein on 18 Feb 41272896 10:13 UTC**
Even though I have a 64-bit machine, I did some testing with the 32-bit install of BOINC.

I compared the stock BOINC install (which uses libcurl 7.19.7) against running BOINC with the newest libcurl dll files.

It looks like the 7.21.4 versions of the libcurl dll files do not really help to alleviate this particular problem. I do, however, think it is worth upgrading to the latest version, especially because of the vulnerabilities that have been patched.

**Comment by romw** [comment]
**Commented by JacobKlein on 31 May 41272925 20:53 UTC**
And I did some further research. I found a man page about CURL located here: http://curl.haxx.se/docs/manpage.html ... and, although there are a bunch of command line options, I'm pretty sure the dll calls are equivalent in design.

My point is that, for the "--limit-rate <speed>" command line switch, the man page states: "The given rate is the average speed counted during the entire transfer.  It means that curl might use higher transfer speeds in short bursts, but  over time it uses no more than the given rate."

So, I suppose that this ticket is a limitation with libcurl, and I'll remark it as wontfix. But I still think we should consider upgrading our libcurl libraries, and possibly try to keep them up to date.
